### PR TITLE
[SPARK-33282] Migrate from deprecated probot autolabeler to GitHub labeler action

### PR DESCRIPTION
**Pull request by kbendick**
### What changes were proposed in this pull request?

This PR removes the old Probot Autolabeler labeling configuration, as the probot autolabeler has been deprecated. I've updated the configs in Iceberg and in Avro, and we also need to update here. This PR adds in an additional workflow for labeling PRs and migrates the old probot config to the new format. Unfortunately, because certain features have not been released upstream, we will not get the _exact_ behavior as before. I have documented where that is and what changes are neeeded, and in the associated ticket I've also discussed other options and why I think this is the best way to go. Definitely a follow up ticket is needed to get the original behavior back in these few cases, but PRs have not been labeled for almost a month and so it's probably best to get it right 95% of the time and occasionally have some UI related PRs labeled as `CORE` while the issue is resolved upstream and/or further investigated. 

### Why are the changes needed?

The probot autolabeler is dead and will not be maintained going forward. This has been confirmed with github user [at]mithro in an issue in their repository.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

To test this PR, I first merged the config into my local fork. I then edited it several times and ran tests on that.

Unfortunately, I've overwritten my fork with the apache repo in order to create a proper PR. However, I've also added the config for the same thing in the Iceberg repo as well as the Avro repo.

I have now merged this PR into my local repo and will be running some tests on edge cases there and for validating in general:
- [Check that the SQL label is applied for changes directly below repo root's sql directory](https://github.com/kbendick/spark/pull/16) ✅ 
- [Check that the structured streaming label is applied](https://github.com/kbendick/spark/pull/20) ✅ 
- [Check that a wildcard at the end of a pattern will match nested files](https://github.com/kbendick/spark/pull/19) ✅ 
- [Check that the rule **/*pom.xml will match the root pom.xml file](https://github.com/kbendick/spark/pull/25) ✅ 

I've also discovered that we're likely not killing github actions that run (like large tests etc) when users push to their PR. In most cases, I see that a user has to mark something as "OK to test", but it still seems like we might want to discuss whether or not we should add a cancellation step In order to save time / capacity on the runners. If so desired, we would add an action in each workflow that cancels old runs when a `push` action occurs on a PR. This will likely make waiting for test runners much faster iff tests are automatically rerun on push by anybody (such as PMCs, PRs that have been marked OK to test, etc). We could free a large number of resources potentially if a cancellation step was added to all of the workflows in the Apache account (as github action API limits are set at the account level).

Admittedly, the fact that the "old" workflow runs weren't cancelled could admittedly be because of the fact that I was working in a fork, but given that there are explicit actions to be added to the start of workflows to cancel old PR workflows and given that we don't have them configured indicates to me that likely this is the case in this repo (and in most `apache` repos as well), at least under certain circumstances (e.g. repos that don't have "Ok to test"-like webhooks as one example).

This is a separate issue though, which I can bring up on the mailing list once I'm done with this PR. Unfortunately I've been very busy the past two weeks, but if somebody else wanted to work on that I would be happy to support with any knowledge I have.

The last Apache repo to still have the probot autolabeler in it is Beam, at which point we can have Gavin from ASF Infra remove the permissions for the probot autolabeler entirely. See the associated JIRA ticket for the links to other tickets, like the one for ASF Infra to remove the dead probot autolabeler's read and write permissions to our PRs in the Apache organization.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
ok to test

File: .github/labeler.yml
Comment (kbendick) [inline .github/labeler.yml:117]: To test for possible matches, I've been using `find` and `grep`. For this one, I believe we likely do need an initial glob as I found the following matches. I will add another test against one of these files.

```bash
$ find . -type d -iname 'streaming' | grep sql
./python/test_support/sql/streaming
./examples/src/main/python/sql/streaming
./examples/src/main/java/org/apache/spark/examples/sql/streaming
./examples/src/main/scala/org/apache/spark/examples/sql/streaming
./sql/core/src/test/java/test/org/apache/spark/sql/streaming
./sql/core/src/test/scala/org/apache/spark/sql/streaming
./sql/core/src/test/scala/org/apache/spark/sql/execution/streaming
./sql/core/src/main/java/org/apache/spark/sql/streaming
./sql/core/src/main/java/org/apache/spark/sql/execution/streaming
./sql/core/src/main/scala/org/apache/spark/sql/streaming
./sql/core/src/main/scala/org/apache/spark/sql/execution/streaming
./sql/core/src/main/scala-2.13/org/apache/spark/sql/execution/streaming
./sql/core/src/main/scala-2.12/org/apache/spark/sql/execution/streaming
./sql/catalyst/src/test/java/org/apache/spark/sql/streaming
./sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/streaming
./sql/catalyst/src/main/java/org/apache/spark/sql/streaming
./sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/streaming
./sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/streaming
./sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/streaming
```

Comment (kbendick) [reply .github/labeler.yml:117]: As demonstrated in this PR, a leading `**` is needed to match the entries that are not in the root sql directory. https://github.com/kbendick/spark/pull/23

Comment (kbendick) [reply .github/labeler.yml:117]: I'm pushing a change for this as I feel that changes to `./examples/src/main/scala/org/apache/spark/examples/sql/streaming/...` should be labeled as STRUCTURED STREAMING. It also matches the original configuration better in my opnion.

File: .github/labeler.yml
Comment (kbendick) [inline .github/labeler.yml:67]: Along the same lines as the `STRUCTURED STREAMING` tag, I'm wondering if this one also might need a leading glob for subdirectories called core.

Comment (kbendick) [reply .github/labeler.yml:67]: There are only three possible matches that I can find, as listed below. There are only three and of those one of them is the root `core` directory. So let's look at the other two.

```bash
$ find . -type d -iname 'core'           
./resource-managers/kubernetes/core
./core
./sql/core
```

Looking at the files in /resource-managers/kubernetes/core, it doesn't feel like any of them are "core" - but I could be misinformed on what we consider `CORE`. The files underneath the kubernetes path are all related to k8s deployment and scheduling (and all of the things that go with that on k8s, such as persistent volume claims and persistent volume mounts, etc). While these do feel potentially like `core` behaviors, if this were the case I would expect the other cluster managers to also fall under this section.

As for `./sql/core`, this one I do believe should be a match. One of the files below it is `sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala`. Definitely I think that the main file defining the `Dataset` type is core.

I would propose that we add `sql/core/**/*` to the list of globs for `CORE` so as to match everything underneath `sql/core` but to still avoid the things underneath `resource-managers`.

File: .github/labeler.yml
Comment (kbendick) [inline .github/labeler.yml:59]: I've also found some `README` files that don't end in `.md`. Not sure if these should be updated to use `.md` or if we should add a match for `**/README*`.

These are the two files:
```
- ./assembly/README
- ./mllib/src/main/resources/org/apache/spark/ml/feature/stopwords/README
```

File: .github/labeler.yml
Comment (kbendick) [reply .github/labeler.yml]: This has been updated to `"**/sql/**/streaming/**/*"` and the TODO comment has been removed.

File: .github/labeler.yml
Comment (kbendick) [reply .github/labeler.yml:67]: I've yet to take action on this one, but I'm leading towards adding an additional rule for `sql/core/**/*` so as to avoid the resource-managers/kubernetes/core folders while still picking up the "core" SQL files such as Dataset.scala etc.

Does anybody have any thoughts on whether or not we'd like the files underneath `./resource-managers/kubernetes/core` to be labeled as `CORE`? My gut tells me no, but I'm still pretty new here so I'll go with whatever you guys suggest.  cc @HyukjinKwon @dongjoon-hyun @RussellSpitzer @rdblue 

After that, I believe this PR is ready to be merged. We will need to open an issue for the `actions/labeler` team to cut a release with `any` / `all` support. Their current suggestion in a relatively old issue (linked in the main issue at https://issues.apache.org/jira/browse/SPARK-33282) is to use the `main` branch (renamed from `master`), but I had difficulty with getting matches using that and overall I'm not that excited about the idea of using a non-released version.

I'd prefer to open an issue to cut a release with support for `all` and `any` (we really only need `any` to handle the negated matches correctly), add a link to the issue to the labeler file, and then merge this and update it once said release has been cut.

**Comment by kbendick** [comment]
I created an issue in the `actions/labeler` repo about releasing a version with proper support for `any` and `all`, which would allow us to use the negation (`!`) operator on matches and get the correct behavior again: https://github.com/actions/labeler/issues/111

File: .github/labeler.yml
Comment (kbendick) [inline .github/labeler.yml:87]: This is how the `any` should look when it's supported in a release upstream.

**Comment by HyukjinKwon** [comment]
Thanks for working on this @kbendick. Let me know when you think it's ready to merge. We can merge and see how it gose.

File: .github/labeler.yml
Comment (HyukjinKwon) [reply .github/labeler.yml:75]: I intended only to have a match with the root `core` when I added. If it works in a different way, we should fix it.
I would prefer to keep `sql/core/**/*` as `SQL`. That's what I usually have seen from other PRs.

Also, let's focus on converting it to labeler action instead of adding/removing/changing other patterns here.

File: .github/labeler.yml
Comment (HyukjinKwon) [reply .github/labeler.yml]: SGTM

File: .github/labeler.yml
Comment (HyukjinKwon) [reply .github/labeler.yml:67]: SGTM

File: .github/labeler.yml
Comment (HyukjinKwon) [inline .github/labeler.yml:60]: We can either wait for them to support in this PR, or merge this PR first with a followup later when they start to support. I don't mind either way - whichever way you prefer.

**Comment by HyukjinKwon** [comment]
cc @nchammas FYI

File: .github/labeler.yml
Comment (kbendick) [reply .github/labeler.yml:60]: Cool. I opened an issue upstream asking when they thought they might be releasing a stable build with `any` / `all`. We can follow up on them later so that at least the large majority of PRs will be correctly labeled.

File: .github/labeler.yml
Comment (kbendick) [reply .github/labeler.yml:67]: Ok. I'll leave them as is. I'm not sure what exactly their purpose is so I'll follow up in another PR on those two files.

File: .github/labeler.yml
Comment (kbendick) [reply .github/labeler.yml:75]: Ah ok. I wasn't sure on the original intended behavior for a few of them but I agree let's keep it focused on converting it to the labeler action for now (other than the few bits that don't seem to be stable yet - though I'm going to give those another go now that I've gotten a lot more experience with this labeler).

TLDR: I will leave it as originally intended 👍

**Comment by kbendick** [comment]
> Thanks for working on this @kbendick. Let me know when you think it's ready to merge. We can merge and see how it gose.

Thanks @HyukjinKwon! This is ready to be merged.

It might need a small tweak or two given it was a pretty complicated config relative to the others that I've done so far but I would say it's tested and ready to merge so people can finally have their labels back. 👍

**Comment by HyukjinKwon** [comment]
Let me merge and see how it goes.

Thanks for this work and thanks for the followup in advance. BTW would you mind file a separate JIRA to follow up?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: .github/labeler.yml
Comment (kbendick) [reply .github/labeler.yml:60]: The issue is mentioned in the JIRA issue I opened up about following up on the labeler.

**Comment by kbendick** [comment]
I created an issue for updating the labeler with `any` to match the exact behavior as before. It also links to the issue I opened in the `actions/labeler` repo asking if they have a potential timeline for a release that contains the any function. It's up to you if you want to use HEAD of their main branch, which does have that support cc @HyukjinKwon @dongjoon-hyun  https://issues.apache.org/jira/browse/SPARK-33370

**Comment by HyukjinKwon** [comment]
Just a reminder :-). It might be great to double check if `any` match is supported now.

**Pull request closed without merging**
